singularity-eos/eos/eos_stiff.hpp

```diff
@@ -0,0 +1,200 @@
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#ifndef _SINGULARITY_EOS_EOS_EOS_STIFF_HPP_
+#define _SINGULARITY_EOS_EOS_EOS_STIFF_HPP_
+
+// stdlib
+#include <cmath>
+#include <cstdio>
+#include <string>
+
+// Ports-of-call
+#include <ports-of-call/portability.hpp>
+
+// Base stuff
+#include <singularity-eos/base/constants.hpp>
+#include <singularity-eos/base/eos_error.hpp>
+#include <singularity-eos/base/robust_utils.hpp>
+#include <singularity-eos/eos/eos_base.hpp>
+
+#define MYMAX(a, b) a > b ? a : b
+
+namespace singularity {
+
+using namespace eos_base;
+
+class StiffGas : public EosBase<StiffGas> {
+ public:
+  StiffGas() = default;
+  PORTABLE_INLINE_FUNCTION StiffGas( Real gm1,
+                                     Real Cv,
+                                     Real Pinf,
+                                     Real qq,
+                                     Real qp )
+      : _Cv(Cv), _gm1(gm1), _Pinf(Pinf), _qq(qq), _qp(qp),
+        _rho0( (_P0 + _Pinf) / (_gm1*_Cv*_T0) ),
+        _sie0( (_P0 + (_gm1+1.0)*_Pinf)/(_P0 + _Pinf) * _Cv*_T0 + _qq ),
+        _entr0( _Cv * log(_T0/( pow(_rho0*_gm1*_Cv, _gm1) ) + _qp )),
+        _bmod0( pow(_rho0,_gm1)*pow(_Cv,_gm1+1.0)*pow(_gm1,_gm1+2.0)*exp((_entr0-_qp)/_Cv) ),
+        _dpde0( _rho0*_gm1 ) {
+    checkParams();
+  }
+  StiffGas GetOnDevice() { return *this; }
+  PORTABLE_INLINE_FUNCTION Real TemperatureFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    return MYMAX( 0.0, (rho*(sie-_qq)-_Pinf) / (rho*_Cv) );
+  }
+  PORTABLE_INLINE_FUNCTION void checkParams() const {
+    // Portable_require seems to do the opposite of what it should. Conditions
+    // reflect this and the code should be changed when ports-of-call changes
+    PORTABLE_ALWAYS_REQUIRE(_Cv >= 0, "Heat capacity must be positive");
+    PORTABLE_ALWAYS_REQUIRE(_gm1 >= 0, "Gruneisen parameter must be positive");
+  }
+  PORTABLE_INLINE_FUNCTION Real InternalEnergyFromDensityTemperature(
+      const Real rho, const Real temperature, Real *lambda = nullptr) const {
+    return MYMAX( 0.0, (rho*_Cv*temperature+_Pinf)/rho + _qq );
+  }
+  PORTABLE_INLINE_FUNCTION Real PressureFromDensityTemperature(
+      const Real rho, const Real temperature, Real *lambda = nullptr) const {
+    return MYMAX(0.0, _gm1 * rho * _Cv * temperature - _Pinf );
+  }
+  PORTABLE_INLINE_FUNCTION Real PressureFromDensityInternalEnergy(
+      const Real rho, const Real sie, Real *lambda = nullptr) const {
+    return MYMAX(0.0, _gm1 * rho * (sie-_qq) - (_gm1+1.0)*_Pinf );
+  }
+  PORTABLE_INLINE_FUNCTION Real EntropyFromDensityTemperature(
+      const Real rho, const Real temperature, Real *lambda = nullptr) const {
+    return _Cv * log( temperature / (pow(rho*_gm1*_Cv,_gm1)) ) + _qp;
```

Comment:
If there's any chance the denominator might go to zero even numerically, you should use `robust::ratio` instead of division.

Similarly, you may also wish to guard against `log(0)`. 

Also in C++ it's good practice to write `std::log` instead of `log`, though it's not a big deal either way.

Reply:
I see. Should I also use `std::pow` ? I'm not sure what the convention normally is in C++ on a lot of things. How would you recommend guarding against `log(0)`

Reply:
Yeah `std::pow` over `pow` is good too. For guarding against `log(0)` it's enough to add a small number to the value you're taking the log of so it's never exactly zero.

Reply:
One thing you'll learn about me is that I am not a fan of pow. I'd actually recommend trying replacing `log( temperature / (pow(rho*_gm1*_Cv,_gm1)) )` with `log( temperature) - _gm1* log(rho*_gm1*_Cv)`, 2 logs is likely much cheaper than 1 log and 1 pow. 

I'll be going through these expressions and making many similar suggestions. Please don't take it as criticism @c0sm0-kramer, I kinda like doing it and it ensures better performance.

Reply:
Before we move too much further on this, we should ask _how_ we want to calculate entropy. If you use the integrals [here](https://lanl.github.io/singularity-eos/main/src/models.html#entropy-availability), you need to specify a reference density and temperature at which the entropy is defined to be zero.

This is how we have done it for the [ideal gas EOS](https://lanl.github.io/singularity-eos/main/src/models.html#ideal-gas), and actually the stiffened gas has the exact same entropy formula. My preference is towards defining a reference density and temperature rather than a reference entropy myself since the former are more meaningful.

I'll also add that you don't need the `pow` function if you use the formulation I linked above. Also you will need to specify a reference temperature and density in the input, but you won't have to specify `qp`.

Reply:
Just want to document here from other conversations with @c0sm0-kramer that a reference entropy might still be useful as well. I think it's just more useful to pair that with a specified reference state than to have a value that isn't as clearly tied to a reference state. We're still iterating on what this should look like

Reply:
Entropy converted to a more general format like in the ideal gas case

Reply:
Additional input is left to reference pressure/temperature and an entropy offset.